### PR TITLE
[Progression] Implement Match Goals menu and XP tracking

### PR DESCRIPTION
**Pull request by Zanieon**
Currently Match Goals menu is unused and will never update properly regardless if player is earning XP through matches, for the post summary screen all of that will count as Double XP due to improper leveling tracking of where each XP point is going into, so this PR should fully implements and integrates `_challenges.gnut` script, solving this problem.

- [x] Implement Weapon Leveling tracking
- [x] Implement Titan Leveling tracking
- [x] Implement Match Score for all vanilla gamemodes
- [x] Implement Match Complete bonus
- [x] Implement Match Won bonus
- [x] Implement Successful Evac bonus
- [x] Implement Squad Leader bonus
- [x] Implement proper Double XP tracking
- [x] Implement Faction Leveling tracking (invisible in esc menu, but counts in post-summary)
- [x] Properly Rewards Advocate Gifts for leveling anything

Clipboard01

### CODE REVIEW

File: Northstar.CustomServers/mod/scripts/vscripts/_xp.gnut [hunk 1]
```diff
@@ -30,11 +33,19 @@ bool function PlayerProgressionAllowed( entity player )
 void function HandleXPGainForScoreEvent( entity player, ScoreEvent event )
 {
 	// note: obviously all xp stuff can be cheated in if people want to on customs, this is mainly just here for fun for those who want it and feature completeness
-
+
```
Comment (GeckoEidechse) [inline Northstar.CustomServers/mod/scripts/vscripts/_xp.gnut:36]: Unrelated whitespace change

Comment (Zanieon) [reply Northstar.CustomServers/mod/scripts/vscripts/_xp.gnut]: Notepad++ auto-apply a Tab space, solved.

**Comment by Zanieon** [comment]
All of vanilla gamemodes have proper hooks for their Merit Score Challenges now, shall i expand that into custom ones located in `Northstar.Custom`?

**Review by ASpoonPlaysGames — APPROVED** [review]
Code looks good at a glance

**Comment by AshesToAshes76** [comment]
Not able to test the code at the moment, but having a better/working xP + merit system would greatly improve the Vanilla experience and give players who want to have Progression enabled more options to permanently unlock weapons, titans, and skins.

**Comment by Zanieon** [comment]
Can we please get this merged, i have other stuff to PR which i considers the implementation of this PR on them.

**Comment by GeckoEidechse** [comment]
I'm assuming you tested this a bunch on your FD servers? ^^

**Comment by Zanieon** [comment]
I integrated this already in my FD branch, so all FD servers being hosted right now shows this PR working.

**Review by Klemmbaustein — APPROVED** [review]
I had some weird issues when first testing thist, but after checking out the PR again it worked fine.

Match goals and XP tracking are working and I don't see anything wrong with the code.

**Comment by Jan200101** [comment]
Tested on a local listen server
Appears to do what it says on the tin
image

I initially tried testing it with attrition but FlightCore installs PRs without installing a navmesh so I had to manually spawn them in

**Review by Jan200101 — APPROVED** [review]
Code looks good overall and follows established patterns.
Lots of it is just checking if a goal has been completed. and informing the client accordingly.

**Comment by Jan200101** [comment]
unsure if any more feedback is needed @GeckoEidechse

**Comment by GeckoEidechse** [comment]
Thanks so much for the review and the ping <3

**Comment by GeckoEidechse** [comment]
Merging based on reviews o7
